### ISSUE TITLE
Table not created

### ISSUE THREAD

**Issue by kttary** [ngageoint/geopackage-android#48]
I'm trying to create a database and table from tab fragment. My problem is the table is not created. Should database creation run on MainActivity or there is something i forgot in my code? Thank you

#### Version Information:

  * GeoPackage Android Version: mil.nga.geopackage.map:geopackage-android-map:3.0.2
  * GeoPackage Android Source: Central Repository 
  * Device or Emulator: AVD

#### Expected Results:

  * The table TESTFEATURE created

#### Observed Results:

  * the database created but the table does not

#### Output:

  * No error

#### Steps to Reproduce:

  1. Create a small app with two tab. Tab 1 contain map and the second contains button
  2. Click the button
  3. Cek the database created in /databases folder

#### Relevant Code:

```
btnCreateDB.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                GeoPackageManager manager = GeoPackageFactory.getManager(getActivity());

                manager.delete("test_gpkg.gpkg");
                try {
                    manager.create("test_gpkg.gpkg");

                    GeoPackage geoPackage = manager.open("test_gpkg.gpkg");
                    if (geoPackage == null) {
                        Log.d("Error null", "Error geopackage null");
                    }
                    geoPackage.createGeometryColumnsTable();

                    SpatialReferenceSystemDao srsDao = geoPackage.getSpatialReferenceSystemDao();
                    SpatialReferenceSystem epsgSrs = srsDao.queryForId(4326l);
                    srsDao.createWebMercator();

                    GeometryColumnsDao geometryColumnsDao = geoPackage.getGeometryColumnsDao();

                    List<FeatureColumn> columns = new ArrayList<FeatureColumn>();
                    columns.add(FeatureColumn.createPrimaryKeyColumn(0, "ID"));
                    columns.add(FeatureColumn.createGeometryColumn(1, "GEOM",
                            GeometryType.MULTIPOLYGON, false, null));
                    columns.add(FeatureColumn.createColumn(2, "DESC",
                            GeoPackageDataType.TEXT, false, null));

                    FeatureTable table = new FeatureTable("TESTFEATURE", columns);
                    geoPackage.createFeatureTable(table);

                    ContentsDao contentsDao = geoPackage.getContentsDao();
                    Contents contents = new Contents();
                    contents.setTableName("TESTFEATURE");
                    contents.setDataType(ContentsDataType.FEATURES);
                    contents.setIdentifier("TESTFEATURE");
                    contents.setDescription("TEST FEATURE");
                    contents.setMinX(94d);
                    contents.setMinY(-12d);
                    contents.setMaxX(142d);
                    contents.setMaxY(7d);
                    contents.setSrs(epsgSrs);

                    contentsDao.create(contents);

                    GeometryColumns geometryColumns = new GeometryColumns();
                    geometryColumns.setContents(contents);
                    geometryColumns.setColumnName("GEOM");
                    geometryColumns.setGeometryType(GeometryType.MULTIPOLYGON);
                    geometryColumns.setSrs(contents.getSrs());
                    geometryColumns.setZ((byte) 0);
                    geometryColumns.setM((byte) 0);
                    geometryColumnsDao.create(geometryColumns);

                    Toast toast = Toast.makeText(getContext(), "Table created successfully", Toast.LENGTH_SHORT);
                    toast.setGravity(Gravity.CENTER, 0, 0);
                    toast.show();

                }
                catch (Exception ex)
                {
                    Log.d("Error",  ex.getMessage());
                    Toast toast = Toast.makeText(getContext(), ex.getMessage(), Toast.LENGTH_SHORT);
                    toast.setGravity(Gravity.CENTER, 0, 0);
                    toast.show();
                }
            }
        });
```

**Comment by kttary** [comment]
I was wondering where is the test_gpkg.gpkg located? If i add some features and show the data on top of my map, its show me correctly. but if i save the test_gpkg.gpkg from /user/0/my_app/databases/test_gpkg.gpkg (via device file explorer )
 to desktop, and open it from db browser for sqlite, the table is not listed. Furthermore, qgis dont recognize it as a valid geopackage file.
Why is db bowser unable to list the table? and why is qgis dont recognize it as a valid geopackage?



<img width="948" alt="screen shot 2018-09-07 at 16 31 56" src="https://user-images.githubusercontent.com/39557649/45211105-98fac580-b2bb-11e8-82de-eea9817f4310.png">

**Comment by bosborn** [comment]
I was able to duplicate your scenario and it looks like you are only missing a close statement before copying off of the the device.
``` geoPackage.close(); ```

Yes the path when created internally is something like:
``` /data/user/0/<app_package>/databases/test_gpkg.gpkg ```

I copied the private app database off using
``` adb -d shell "run-as <app_package> cat /data/data/<app_package>/databases/test_gpkg.gpkg" > test_gpkg.gpkg ```
There are also methods to export the GeoPackage to an external device location, or you can perform a share with an intent / copying to cache.

Before adding the geoPackage.close()...
![screen shot 2018-09-07 at 6 51 14 am](https://user-images.githubusercontent.com/11667761/45220197-9e7bef80-b26a-11e8-8f06-fb9bbc77b8b0.png)

After adding the geoPackage.close()...
![screen shot 2018-09-07 at 6 41 28 am](https://user-images.githubusercontent.com/11667761/45220220-ae93cf00-b26a-11e8-9433-90c0e9375286.png)

**Comment by kttary** [comment]
this is the second time i stumbled with geoPackage.close(). I will keep in mind that this method as important as geoPackage.open(). Thank you very much.

**Comment by kttary** [comment]
Today i tried to create Tile table with my classic problem. the problem is that the tile table successfully created and i'm able to open it from sqlite db browser but the file is not recognized as a valid geopackage file from qgis. Actually i just translate from my swift working version with little modification of adding tile taken from [this page](https://github.com/ngageoint/geopackage-java/blob/master/src/test/java/mil/nga/geopackage/test/tiles/UrlTileGeneratorUtils.java). Here is the code

```
@Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_exchange, container, false);
        Button btnCreateDB = (Button) rootView.findViewById(R.id.createdb);

        btnCreateDB.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                GeoPackageManager manager = GeoPackageFactory.getManager(getActivity());
                manager.delete("test_gpkg.gpkg");
                try {
                    manager.create("test_gpkg.gpkg");
                    GeoPackage geoPackage = manager.open("test_gpkg.gpkg");

                    SpatialReferenceSystemDao srsDao = geoPackage.getSpatialReferenceSystemDao();
                    SpatialReferenceSystem epsgSrs = srsDao.queryForId(4326l);
                    srsDao.createWebMercator();

                    geoPackage.createGeometryColumnsTable();

                    GeometryColumnsDao geometryColumnsDao = geoPackage.getGeometryColumnsDao();
                    ContentsDao contentsDao = geoPackage.getContentsDao();

                    createBaseMap(geoPackage, contentsDao, epsgSrs);
                    geoPackage.close();

                }
                catch (Exception ex)
                {
                    Log.d("Error",  ex.getMessage());
                }
            }
        });

        return  rootView;
    }

private void createBaseMap(GeoPackage geoPackage, ContentsDao contentsDao, SpatialReferenceSystem epsgSrs)
    {
        try {
            SpatialReferenceSystemDao srsDao = geoPackage.getSpatialReferenceSystemDao();

            BoundingBox worldBoundingBox = TileBoundingBoxUtils.getWebMercatorBoundingBox(0, 0, 0);
            geoPackage.createTileMatrixSetTable();
            geoPackage.createTileMatrixTable();

            Contents contents = new Contents();
            contents.setTableName("BASEMAP");
            contents.setDataType(ContentsDataType.TILES);
            contents.setIdentifier("BASEMAP");
            contents.setDescription("MY BASE MAP");
            contents.setLastChange(new Date());
            contents.setMinX(worldBoundingBox.getMinLongitude());
            contents.setMinY(worldBoundingBox.getMinLatitude());
            contents.setMaxX(worldBoundingBox.getMaxLongitude());
            contents.setMaxY(worldBoundingBox.getMaxLatitude());
            contents.setSrs(srsDao.queryForId(3857l));

            List<TileColumn> tileColumns = TileTable.createRequiredColumns();
            TileTable tileTable = new TileTable("BASEMAP", tileColumns);
            geoPackage.createTileTable(tileTable);

            contentsDao.create(contents);

            TileMatrixSetDao tileMatrixSetDao = geoPackage.getTileMatrixSetDao();
            TileMatrixSet tileMatrixSet = new TileMatrixSet();
            tileMatrixSet.setContents(contents);
            tileMatrixSet.setSrs(srsDao.queryForId(3857l));
            contents.setMinX(worldBoundingBox.getMinLongitude());
            contents.setMinY(worldBoundingBox.getMinLatitude());
            contents.setMaxX(worldBoundingBox.getMaxLongitude());
            contents.setMaxY(worldBoundingBox.getMaxLatitude());

            tileMatrixSetDao.create(tileMatrixSet);

            insertTile(geoPackage, contents, tileMatrixSet, epsgSrs);
        }
        catch (Exception ex)
        {
            Log.d("Error Tile Table", ex.getMessage());
        }
    }

private void insertTile(GeoPackage geoPackage, Contents contents, TileMatrixSet tileMatrixSet, SpatialReferenceSystem epsgSrs) {

        try {
            Projection projection = epsgSrs.getProjection();

            ProjectionTransform transformTo = projection.getTransformation(3857l);
            ProjectionTransform transformFrom = epsgSrs.getTransformation(transformTo.getToProjection());

            GeometryEnvelope env4326 = new GeometryEnvelope(94d, -11d, 141d, 6d);
            GeometryEnvelope env3857 = transformTo.transform(env4326);

            TileMatrixDao tileMatrixDao = geoPackage.getTileMatrixDao();
            Integer minZoom = 3;
            Integer maxZoom = 5;
            TileGrid totalTileGrid = TileBoundingBoxUtils.getTileGrid(new BoundingBox(env3857.getMinX(), env3857.getMinY(), env3857.getMaxX(), env3857.getMaxY()), minZoom);
            BoundingBox totalBoundingBox = TileBoundingBoxUtils.getWebMercatorBoundingBox(totalTileGrid, minZoom);
            TileGrid tileGrid = totalTileGrid;
            TileDao myDao = geoPackage.getTileDao("BASEMAP");

            for (int zoom = minZoom; zoom < maxZoom + 1; zoom++) {
                byte[] tileData = downloadTile(zoom, tileGrid.getMinX(), tileGrid.getMinY());

                Integer tileWidth = 256;
                Integer tileHeight = 256;
                for (Long x = tileGrid.getMinX(); x < tileGrid.getMaxX() + 1; x++) {
                    for (Long y = tileGrid.getMinY(); y < tileGrid.getMaxY() + 1; y++) {
                        tileData = downloadTile(zoom, tileGrid.getMinX(), tileGrid.getMinY());

                        TileRow newRow = myDao.newRow();
                        newRow.setZoomLevel(zoom);
                        newRow.setTileColumn(x - tileGrid.getMinX());
                        newRow.setTileRow(y - tileGrid.getMinX());
                        newRow.setTileData(tileData);

                        myDao.create(newRow);
                    }
                }

                Long matrixWidth = tileGrid.getMaxX() - tileGrid.getMinX() + 1;
                Long matrixHeight = tileGrid.getMaxY() - tileGrid.getMinY() + 1;
                Double pixelXSize = (tileMatrixSet.getMaxX() - tileMatrixSet.getMinX()) / (matrixWidth * tileWidth);
                Double pixelYSize = (tileMatrixSet.getMaxY() - tileMatrixSet.getMinY()) / (matrixHeight * tileHeight);

                TileMatrix tileMatrix = new TileMatrix();
                tileMatrix.setContents(contents);
                tileMatrix.setZoomLevel(zoom);
                tileMatrix.setMatrixWidth(matrixWidth);
                tileMatrix.setMatrixHeight(matrixHeight);
                tileMatrix.setTileWidth(tileWidth);
                tileMatrix.setTileHeight(tileHeight);
                tileMatrix.setPixelXSize(pixelXSize);
                tileMatrix.setPixelYSize(pixelYSize);
                tileMatrixDao.create(tileMatrix);

                tileGrid = TileBoundingBoxUtils.tileGridZoomIncrease(tileGrid, zoom + 1);
            }
        }
        catch (Exception ex)
        {
            Log.d("Error", ex.getMessage());
        }
    }

    protected byte[] downloadTile(int z, long x, long y) {

        byte[] bytes = null;

        String zoomUrl = String.format("https://osm.geointservices.io/osm_tiles/%1$s/%2$s/%3$s.png",String.valueOf(z), String.valueOf(x), String.valueOf(y));

        zoomUrl = URLDecoder.decode(zoomUrl);
        URL url;
        try {
            url = new URL(zoomUrl);
        } catch (Exception ex)
        {
            throw new GeoPackageException("Failed to download tile. URL: " + zoomUrl + ", z=" + z + ", x=" + x + ", y=" + y, ex);
        }

        HttpURLConnection connection = null;
        try {
            connection = (HttpURLConnection) url.openConnection();
            connection.connect();

            int responseCode = connection.getResponseCode();
            Log.d("test", String.valueOf(connection.getResponseCode()));

            if (responseCode == HttpURLConnection.HTTP_MOVED_PERM
                    || responseCode == HttpURLConnection.HTTP_MOVED_TEMP
                    || responseCode == HttpURLConnection.HTTP_SEE_OTHER) {
                String redirect = connection.getHeaderField("Location");
                connection.disconnect();
                url = new URL(redirect);
                connection = (HttpURLConnection) url.openConnection();
                connection.connect();
            }

            if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
                throw new GeoPackageException("Failed to download tile. URL: "
                        + zoomUrl + ", z=" + z + ", x=" + x + ", y=" + y);
            }

            InputStream geoPackageStream = connection.getInputStream();
            bytes = GeoPackageIOUtils.streamBytes(geoPackageStream);

        } catch (IOException e) {
            throw new GeoPackageException("Failed to download tile. URL: "
                    + zoomUrl + ", z=" + z + ", x=" + x + ", y=" + y, e);
        } finally {
            if (connection != null) {
                connection.disconnect();
            }
        }

        return bytes;
    }
```

Here is error in qgis 
<img width="1128" alt="screen shot 2018-09-09 at 12 50 28" src="https://user-images.githubusercontent.com/39557649/45261523-05e29c80-b42f-11e8-8087-e508a57c9286.png">


Thank you

**Comment by bosborn** [comment]
Can you attach or link your GeoPackage?

**Comment by kttary** [comment]
Attached my geopackage contains tilefeature only. It can be opened by sqlite db browser but qgis.

https://www.dropbox.com/s/ita013u88gf47qk/test_gpkg.gpkg?dl=0

Thank you

**Comment by bosborn** [comment]
There are at least a few errors with the code.  The end GeoPackage result does not contain any tile matrix or tile rows.

The first issue appears to be performing network operations on the main thread.  Downloading tiles would need to be done off the main thread, such as in an AsyncTask.

I changed your downloadTile method to return empty bytes to see if that was the only issue.  This resulted in tile rows (w/ empty tiles) but no tile matrix rows.

The second issue looks like the min and max x/y bounds are being set on the Contents once again, and not the TileMatrixSet.  So later when pixel sizes are calculated, they are set to 0 which is invalid.

The last issue I noticed is the call to tileGridZoomIncrease is increasing the zoom level by zoom + 1 when it should be just 1.

After those changes, the resulting GeoPackage contains a Tile Matrix Set row, Tile Matrix rows, and Tile rows (empty images).  The GeoPackage opens in QGIS 3.

You probably want to verify you are not catching, logging, and continuing with errors before exporting the GeoPackage.

**Comment by kttary** [comment]
thank you very much. it getting fixed now. I learn a lot today. However, i'm facing problem when showing featuretile on fragment now. I think previously i have [similar problem](https://github.com/ngageoint/geopackage-android-map/issues/10). It was solved by put geopackage.close() on maploaded events. But now, i have tab fragment. First fragment contain map and the second contain data preparation (including create database, import tiles, etc). On the map tab, i have a button to load database onto the map. the button is set visible on mapready event. if i commenting out geopackage.close() i can see the featuretiles. But i believe it is unsafe. my question is where to put geopackage.close() command? thank you

**Comment by bosborn** [comment]
A GeoPackage is nothing more that a SQLite database file.  So treat it like you would an active database connection.  The connection needs to remain open while you are reading from or writing to it.  This includes if you have a tile overlay adding images to the map.  When the data is no longer "active", the app is closed, or the GeoPackage is going to be exported, you'd want to close the GeoPackage / database connection.

**Comment by bosborn** [comment]
You can use the [GeoPackageCache](https://github.com/ngageoint/geopackage-android/blob/master/geopackage-sdk/src/main/java/mil/nga/geopackage/GeoPackageCache.java) (and [GeoPackageCoreCache](https://github.com/ngageoint/geopackage-core-java/blob/master/src/main/java/mil/nga/geopackage/GeoPackageCoreCache.java)) as an easy way to maintain open connections and close them later.

**Comment by kttary** [comment]
i put it at the end of code but doesn't work

```
GeoPackageManager manager = GeoPackageFactory.getManager(getActivity());
geoPackage = manager.open("test_gpkg.gpkg");

List<String> tileTables = geoPackage.getTileTables();
TileDao tileDao = geoPackage.getTileDao(tileTables.get(tileTables.indexOf("PETADASAR")));
TileProvider overlay = GeoPackageOverlayFactory.getTileProvider(tileDao);
final TileOverlayOptions overlayOptions = new TileOverlayOptions();
overlayOptions.tileProvider(overlay);
overlayOptions.zIndex(-2.0f);
gMap.addTileOverlay(overlayOptions);
geoPackage.close();
```

**Comment by bosborn** [comment]
Right... closing a database connection there will prevent the tile overlay from reading the tiles from the closed database.

**Comment by kttary** [comment]
is there an event that we can catch when reading tiles from database is complete?

**Comment by bosborn** [comment]
It would be "complete" only when the user is done looking at the tiles on the map.  When the map is scrolled, zoomed, etc, new tiles are read from the GeoPackage which needs to remain open.  This is an app specific decision.

You need to maintain an open GeoPackage as long as you expect to get tiles from it on the map in a tile overlay.
You can check out [MapCache](https://github.com/ngageoint/geopackage-mapcache-android) as an example of this.

Basically maintain a collection of open GeoPackages in static memory or using the Cache I mentioned above while they are being viewed on the map.  When the data is no longer needed, app is closed, or you plan to manually copy the database off of the phone, you should close them.

**Comment by kttary** [comment]
i tried to modify my app so then user need to open and close geopackage on their own needs. So the geopackage is never closed until the user close it. But in the middle of activity, android often complain about database is not closed but at the end it fix again.

D/DragAndDrop: Dragging
W/System: A resource failed to call close. 
W/SQLiteConnectionPool: A SQLiteConnection object for database '/data/user/0/id.co.meetbit.dev.petakerja/databases/geopackage_metadata' was leaked!  Please fix your application to end transactions in progress properly and to close the database when it is no longer needed.
W/System: A resource failed to call close. 
          A resource failed to call close. 
W/SQLiteConnectionPool: A SQLiteConnection object for database '/data/user/0/id.co.meetbit.dev.petakerja/databases/geopackage_metadata' was leaked!  Please fix your application to end transactions in progress properly and to close the database when it is no longer needed.
W/System: A resource failed to call close. 
W/System: A resource failed to call close. 
W/SQLiteConnectionPool: A SQLiteConnection object for database '/data/user/0/id.co.meetbit.dev.petakerja/databases/geopackage_metadata' was leaked!  Please fix your application to end transactions in progress properly and to close the database when it is no longer needed.
W/System: A resource failed to call close. 
W/System: A resource failed to call close. 
W/SQLiteConnectionPool: A SQLiteConnection object for database '/data/user/0/id.co.meetbit.dev.petakerja/databases/geopackage_metadata' was leaked!  Please fix your application to end transactions in progress properly and to close the database when it is no longer needed.
W/System: A resource failed to call close. 
W/System: A resource failed to call close. 
W/SQLiteConnectionPool: A SQLiteConnection object for database '/data/user/0/id.co.meetbit.dev.petakerja/databases/geopackage_metadata' was leaked!  Please fix your application to end transactions in progress properly and to close the database when it is no longer needed.
W/System: A resource failed to call close. 
W/System: A resource failed to call close. 
W/SQLiteConnectionPool: A SQLiteConnection object for database '/data/user/0/id.co.meetbit.dev.petakerja/databases/geopackage_metadata' was leaked!  Please fix your application to end transactions in progress properly and to close the database when it is no longer needed.
W/System: A resource failed to call close. 
W/System: A resource failed to call close. 
W/SQLiteConnectionPool: A SQLiteConnection object for database '/data/user/0/id.co.meetbit.dev.petakerja/databases/geopackage_metadata' was leaked!  Please fix your application to end transactions in progress properly and to close the database when it is no longer needed.
W/System: A resource failed to call close. 
W/System: A resource failed to call close. 
W/SQLiteConnectionPool: A SQLiteConnection object for database '/data/user/0/id.co.meetbit.dev.petakerja/databases/geopackage_metadata' was leaked!  Please fix your application to end transactions in progress properly and to close the database when it is no longer needed.
W/System: A resource failed to call close. 
W/System: A resource failed to call close. 
W/SQLiteConnectionPool: A SQLiteConnection object for database '/data/user/0/id.co.meetbit.dev.petakerja/databases/geopackage_metadata' was leaked!  Please fix your application to end transactions in progress properly and to close the database when it is no longer needed.
W/System: A resource failed to call close. 
W/System: A resource failed to call close. 
W/SQLiteConnectionPool: A SQLiteConnection object for database '/data/user/0/id.co.meetbit.dev.petakerja/databases/geopackage_metadata' was leaked!  Please fix your application to end transactions in progress properly and to close the database when it is no longer needed.
W/System: A resource failed to call close. 
W/System: A resource failed to call close. 
W/SQLiteConnectionPool: A SQLiteConnection object for database '/data/user/0/id.co.meetbit.dev.petakerja/databases/geopackage_metadata' was leaked!  Please fix your application to end transactions in progress properly and to close the database when it is no longer needed.
W/System: A resource failed to call close. 
I/chatty: uid=10091(id.co.meetbit.dev.petakerja) FinalizerDaemon identical 11 lines
W/System: A resource failed to call close. 
D/TESST: 1

What i'm doing during the session is doing querying feature which is FeatureIndexResults myRes  = myQuery.queryFeatures(myBB); myRes.close(); What is other object need to close? thank you
